sw/device/lib/testing/test_coverage.c

```diff
@@ -0,0 +1,84 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/uart.h"
+#include "sw/vendor/llvm_clang_rt_profile/compiler-rt/lib/profile/InstrProfiling.h"
+#include <stdint.h>
+
+/**
+ * When the linker finds a definition of this symbol, it knows to skip loading
+ * the object which contains the profiling runtime’s static initializer.
+ */
+int __llvm_profile_runtime;
+
+/**
+ * Start and end of `.init_array` section where the LLVM profiling runtime
+ * initialization function resides, defined in `flash_link.ld`.
+ */
+extern void (*_init_array_start[])(void);
+extern void (*_init_array_end[])(void);
+
+/**
+ * Computes the CRC32 of a buffer as expected by Python's `zlib.crc32()`. The
+ * implementation below is basically a simplified, i.e. byte-by-byte and without
+ * a lookup table, version of zlib's crc32. See
+ * https://github.com/madler/zlib/blob/2fa463bacfff79181df1a5270fb67cc679a53e71/crc32.c,
+ * lines 111-112 and 276-279.
+ */
+static uint32_t crc32(char *buf, size_t len) {
+  // CRC32 polynomial.
+  static const uint32_t kPoly = 0xEDB88320;
+  // Since we use a contiguous buffer, we don't need to call this function
+  // multiple times. That's why `crc` is not a parameter and is initialized to
+  // zero.
+  uint32_t crc = 0;
+  crc = ~crc;
+  for (; len > 0; --len, ++buf) {
+    crc = crc ^ *buf;
+    for (uint8_t i = 0; i < 8; ++i) {
+      if (crc & 1) {
+        crc = kPoly ^ (crc >> 1);
+      } else {
+        crc = crc >> 1;
+      }
+    }
+  }
+  return ~crc;
+}
+
+/**
+ * Sends the given buffer as a hex string over UART.
+ */
+static void send_buffer(char *buf, size_t len) {
+  for (uint32_t i = 0; i < len; ++i) {
+    base_printf("%2X", buf[i]);
+  }
+}
+
+void test_coverage_init_buffer() {
+  size_t cnt = _init_array_end - _init_array_start;
+  for (size_t i = 0; i < cnt; ++i) {
+    _init_array_start[i]();
+  }
+}
+
+void test_coverage_send_buffer() {
+  // It looks like we don't have a way to read the profile buffer incrementally.
+  // Thus, we define the following buffer.
+  char buf[1024] = {0};
```

Comment:
Is this going to scale to actual programs that we want to get coverage from?

Reply:
That is an open question but I have one datapoint :) The length of the profile buffer for `coverage_test.c` is 312 bytes.

Reply:
I tried running `sha256_test` with coverage (3 TUs with coverage: `sha256_test.c`, `hw_sha256.c`, and `hmac.c`) and the buffer size was 856 bytes. There is some fixed cost per file and function but AFAICT the buffer size mostly depends on the number of [mapping regions](https://github.com/lowRISC/opentitan/pull/3080#discussion_r467032508). It looks like we won't have an issue with actual programs but nevertheless I changed the size of the buffer to 8KB to be safe (for reference, our RAM is 64KB according to `top_earlgrey_memory.ld`).